### ISSUE TITLE
Start to transition all Python scripts to JavaScript

### ISSUE THREAD

**Issue by ancientstraits** [emscripten-core/emscripten#16597]
I have a feeling this issue has been made in this repo multiple times, so if anyone knows of any active issues similar to this, feel free to reply with those.

I would have made this a pull request, but transitioning all the scripts to JavaScript would be a mammoth task which would take several pull requests, so I am making this pull request to perhaps serve as a page where updates about this transition can be placed.

There are many advantages to transitioning the Python scripts that emscripten uses to JavaScript:
- Less dependencies: Because of the lack of dependency on Python, there would be no worry about Python module issues or having to install a Python interpreter. This means that there would be less issues to deal with, since Python-related issues would not be relevant.
- Consistency: Since many of the scripts (like the library binding scripts) are already written in JavaScript, developers would have a consistent experience maintaining these scripts.
- Somewhat easy to do: Although the Python codebase in this repository is very large, it certainly would not be difficult to transition the codebase, just time-consuming. Both Python and JavaScript have analogous constructs, like classes and objects/dicts.

I think there should be some guidelines for transitioning these scripts, as well:
- Don't use external modules when they are not needed. This will just create more issues in the long run, so only use an external module if Python has one, or if the function is not in the node.js standard library, and would be difficult to implement.
- Maintain the same comments in the JS as in the Python, for the most part.

I have already started rewriting these scripts, specifically the `system/lib/*.py` scripts, which update packages or push their changes. They are simple, and pretty easy to finish, and I will submit a pull request relating to them once I am done with those scripts.

Anyone who knows both JavaScript and Python should feel free to transition the Python scripts. For me so far, it has been a pretty fun task. Those willing to help should reply to this thread about which scripts they will transition.

**Comment by kripken** [comment]
Thanks for opening this issue!

I am personally in favor of this. I think that historically emscripten began when Node.js was not quite mature enough (over a decade ago now...), and so we used a mix of Python and JS. Writing it today, I'd use prefer JS because we need JS anyhow for things like Acorn and other JS ecosystem things, and a single language is simpler.

With that said, other people might feel otherwise so we should have a discussion here before deciding that this is the right direction. I think the main reason we haven't had such a discussion so far is that no one had the time or interest to actually migrate the scripts, but @ancientstraits your post is a good reason to think about it now.

On a more technical level, I'd say most scripts should be straightforward to port, as you said @ancientstraits . The one area that might be trickier is the test suite. I think I'd be fine leaving that as Python, since I'm not sure it's worth the effort to convert as it doesn't reach end users, and also it is a lot of code. But it also calls into other Python code that we might want to convert to JS. So there might be some questions around that, I'm not sure.

Another note, we may want to also consider TypeScript in this context.

**Comment by ancientstraits** [comment]
Right, we will need to eventually consider the test suite, and what we need to do for that. I think we'll just see when we get there.

As for TypeScript, adding it also poses the question of whether the existing JavaScript codebase will be converted as well. TypeScript also requires an additional compilation process to be implemented, meaning that unlike Python or JS scripts, which could be ran without any compilation, TS scripts need to be compiled (or for TS-Node to be on the system). It also needs a `tsconfig.json` file.

To be honest, I also think TypeScript in the project would be a good idea. I just finished writing a `shutil.copy_tree` substitute in JS called `copyDirSync` (Node has a `fs.cpSync` which can copy directories recursively, but only new versions support it). `copyDirSync` has an `opts` argument with an object containing objects in order to substitute the keyword arguments in `shutil.copy_tree`. I think a TypeScript interface for this argument would be a very good idea, because it would be easier to find what options would be needed.

@kripken The only issue I have is that TypeScript needs a compilation step, while Python code can just be run automatically. I also would not know where to put the compiled JS files - putting them in the same directory as the TS files would clutter everything up. Perhaps we could add it as a git hook or something?

If I had to make a choice, I would just do the transition first, then TypeScript later. I just don't have any ideas on how exactly the TS would be implemented.

**Comment by sbc100** [comment]
I think converting to JavaScript at this point would *not* be a step forwards... at best its is a step sidewise and in many ways I could be considered a step backwards (at least for some like myself who is very comfortable working in python).   If we are going to do a re-write I think we would certainly want to target at language with more static types, such as TypeScript. 

In particular, I would advise against starting such a process with the little utility script in `system/lib/*.py` is good place to start because they don't represent code that run by our users and are not representative of the wider codebase.

BTW, regarding the compilation of typescript we have `deno` which is like node by can run TS natively.

**Comment by ancientstraits** [comment]
> I would advise against starting such a process with the little utility script in `system/lib/*.py` is good place to start
@sbc100 So would it be a good idea to start there, or not?

**Comment by sbc100** [comment]
I think there is really only one python tool that all our user run that is `emcc`.   Is there any reason not to target that one from the start?   I suppose you could start with a much smaller tool such as `embuilder` which is would be a little less work.  But that problem with just to one or the other is that we would end up in a state were we have two implementations of all of the common code.

Honestly I'm not sure I buy the argument that the dependency on python is painful enough to warrent a re-write.  Its mostly a solved issue and we even bundle our own python emsdk.    Can you explain more why you think the python dependency is an issue?

**Comment by ancientstraits** [comment]
> I think there is really only one python tool that all our user run that is `emcc`. Is there any reason not to target that one from the start?

Actually, no. I'll get started on that. The only problem I have is that `deno` is still a pretty new thing, which not many people have installed, which would become a dependency of this project. But we could also add a compilation command to the `package.json` to make it run on `node`.

> Can you explain more why you think the python dependency is an issue?

I admit that one big reason I made this issue was so that Emscripten itself could run on a web browser (the only thing stopping it is the Python scripts). However, other demos have been made online which don't use Emscripten (like `clang-wasm` and `cib`), so that reason is not as important now.

Another reason I made the issue is that one day I had some Python module error when I tried to launch `emcc`. However, as you said, I'm pretty sure that has been fixed, and Python is more stable than ever.

Another reason I had is that I thought pretty new coders would be able to do it, making it an easy task. But perhaps we don't need this as an issue after all.

**Comment by sbc100** [comment]
The argument for running the browser is good one, although for now rather neiche.  And also there has a been really good progress on making CPython work better with emscripten these days.  @tiran has been filing and fixing a bunch of bugs there.  So that might be a blocker for running in the web going forward.

Also, even if we did convert all our Python code to JS, its unlikely that we want to spent the time making it web-compatible.  These days the parts of emscripten the run in JS are designed to run under node and not maintained in a web-compatible way.  Making them run on the web and adding yet another testing configuration might not be worth it.    For running emscripten on the web it might be better to somehow run a user-space version JS-on-JS in order to avoid this requirement.   

In other words, if you goal is to run emscirpten's JS code on the web I think just converting to JS running on node won't be enough.. it will but extra constraints on us.

**Comment by ancientstraits** [comment]
Oh right, I remembered that node uses things like `fs` and stuff. It just isn't that easy to run in a browser.

I will just close this issue for now, as it now seems there is no need for it.

**Comment by sbc100** [comment]
Sorry to push back so strongly, @kripken was certainly more on board the myself, and its seems possible that a re-write into *some* language might on the cards.. I'm just not sure its worth it unless we get a static typing out it.

**Comment by sbc100** [comment]
(speaking of static typing we don't necessarily need a re-write to get it: https://github.com/emscripten-core/emscripten/pull/16624)

**Comment by ancientstraits** [comment]
@sbc100 It's all right. Once I re-evaluated my decision, I lost motivation in this. I have other things I need to do, anyway.
